packages/discovery/src/flatten/ParsedFilesManager.ts

```diff
@@ -0,0 +1,371 @@
+    )
+
+    if (matchingContract !== undefined) {
+      return {
+        contract: matchingContract,
+        file,
+      }
+    }
+
+    const matchingImport = findOne(
+      file.importDirectives,
+      (c) => c.importedName === contractName,
+    )
+
+    if (matchingImport !== undefined) {
+      return this.tryFindContract(
+        matchingImport.originalName,
+        this.resolveImportPath(file, matchingImport.absolutePath),
+      )
+    }
+
+    return undefined
+  }
+
+  findFileDeclaringContract(contractName: string): ParsedFile {
+    const matchingFile = findOne(this.files, (f) =>
+      f.contractDeclarations.some((c) => c.name === contractName),
+    )
+    assert(
+      matchingFile !== undefined,
+      `Failed to find file declaring contract ${contractName}`,
+    )
+
+    return matchingFile
+  }
+
+  findContractDeclaration(
+    contractName: string,
+    file?: ParsedFile,
+  ): ContractFilePair {
+    file ??= this.findFileDeclaringContract(contractName)
+
+    const matchingContract = findOne(
+      file.contractDeclarations,
+      (c) => c.name === contractName,
+    )
+    assert(matchingContract !== undefined, 'Contract not found')
+
+    return {
+      contract: matchingContract,
+      file,
+    }
+  }
+
+  resolveImportPath(fromFile: ParsedFile, importPath: string): ParsedFile {
+    const resolvedPath = importPath.startsWith('.')
+      ? posix.join(posix.dirname(fromFile.path), importPath)
+      : importPath
+
+    const matchingFile = findOne(
+      this.files,
+      (f) => posix.normalize(f.path) === posix.normalize(resolvedPath),
+    )
+    assert(
+      matchingFile !== undefined,
+      `File [${fromFile.path}][${resolvedPath}] not found`,
+    )
+
+    return matchingFile
+  }
+}
+
+// Takes a user defined type name such as `MyLibrary.MyStructInLibrary` and
+// returns only the namespace - the part before the dot.
+function extractNamespace(identifier: string): string {
+  const dotIndex = identifier.indexOf('.')
+  if (dotIndex === -1) {
+    return identifier
+  }
+  return identifier.substring(0, dotIndex)
+}
+
+function decodeRemappings(remappingStrings: string[]): Remapping[] {
+  return remappingStrings.map((r) => {
+    const [prefix, target] = r.split('=')
+
+    assert(r.includes('='), 'Invalid remapping, lacking "=" sign.')
+    assert(prefix !== undefined, 'Invalid remapping, missing prefix.')
+    assert(target !== undefined, 'Invalid remapping, missing target.')
+    return { prefix, target }
+  })
+}
+
+function resolveRemappings(path: string, remappings: Remapping[]): string {
+  const matchingRemappings = remappings.filter((r) => path.startsWith(r.prefix))
+  if (matchingRemappings.length > 0) {
+    const longest = matchingRemappings.reduce((a, b) =>
+      a.prefix.length > b.prefix.length ? a : b,
+    )
+
+    const result = posix.join(longest.target, path.slice(longest.prefix.length))
+    return result
+  }
+
+  return path
+}
+
+function findOne<T>(
```

Comment:
why not use `Array.prototype.find`?

Reply:
Because we want to fail if we find more than one.